keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,657 @@
+-->
+
+## Alternatives
+
+### Alternative approaches
+
+The alternatives are designed to solve the limitation for the maximal number of
+pending workloads which is returned in the status.
+
+#### Coarse-grained ordering information per workload in workload status
+
+The idea is to distribute the ordering information among workloads to avoid
+keeping the ordering information centralized, thus avoiding creating objects
+constrained by the etcd limit.
+
+The main complication with distributing the ordering information is that a
+workload admission, or a new workload with a high priority can move the entire
+ordering, warranting update requests to all workloads in the queue. This could
+mean cascades of thousands of requests after such event.
+
+The proposal to control the number of update requests to workloads when a
+workload is admitted or added, is to bucket workload positions. The bucket
+intervals could grow exponentially, allowing for logarithmic number of requests
+needed. With this approach, the number of requests to update workloads is limited
+by the number of buckets, as only the workloads on bucket boundary are updated.
+
+The update requests could be sent by a periodic routine which iterates over the
+cluster queue and triggers workload reconciliation for workloads for which the
+ordering is changed.
+
+Pros:
+- allows to expose the ordering information for all workloads, guaranteeing the
+  user to know its workload position even if it is beyond the top N threshold
+  in the proposed approach.
+
+Cons:
+- it requires a substantial number of requests when a workload is admitted, or
+  a high priority workload is inserted. For example, assuming 1000 workloads,
+  and expotential bucketing with base 2, this is 10 requests.
+- it is not clear if the coarse-grained information would satisfy user
+  expectations. For example, a user may need to wait long to observe reduction
+  of a bucket.
+- an external system which wants to display a pipeline of workloads needs to
+  fetch all workloads. Similarly, as system which wants to list top 10 workloads
+  may need to query all workloads.
+- a natural extension of the mechanism to return ETA in the workload status
+  may also increase the number of requests in a less controlled way.
+
+#### Ordering information per workload in events or metrics
+
+The motivation for this approach is similar as for distributing the information
+in workload statuses. However, it builds on the assumption that update requests
+are more costly than events or metric updates. For example, sending events or
+updating metrics does not trigger a workload reconciliation.
+
+Pros:
+- more lightweight than updating workload status,
+
+Cons:
+- the API based on events or metrics would be less convenient to end users than
+  object-based.
+- probably still requires bucketing, thus inheriting the usability cons related
+  to bucking from the workload status approach.
+
+#### On-demand http endpoint
```

Comment:
Let's add that a possible way to expose the API is an aggregation layer https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/

Reply:
Added a paragraph on this. It seems promising but will require some extra investigation.